src/query/storage/cache/redis_cache.go

```diff
@@ -262,18 +232,155 @@ func WindowGetOrFetch(
 		End:         time.Unix(align_end, 0),
 	}
 
-	res := cache.Get([]*storage.FetchQuery{align_q})
+	res := cache.Get([]*storage.FetchQuery{align_q}, SimpleKeyPrefix)
 	if res[0] == nil {
 		promRes, err := st.FetchProm(ctx, q, fetchOptions)
 		if err == nil {
-			cache.Set([]*storage.FetchQuery{align_q}, []*storage.PromResult{&promRes})
+			cache.Set([]*storage.FetchQuery{align_q}, []*storage.PromResult{&promRes}, SimpleKeyPrefix)
 		}
 
-		cache.logger.Info("cache miss", zap.String("key", keyEncode(align_q)), zap.Int("size", len(promRes.PromResult.Timeseries)))
+		cache.logger.Info("cache miss", zap.String("key", KeyEncode(align_q, SimpleKeyPrefix)), zap.Int("size", len(promRes.PromResult.Timeseries)))
 		cache.cacheMetrics.CacheMetricsMiss(promRes)
 		return promRes, err
 	}
-	cache.logger.Info("cache hit", zap.String("key", keyEncode(align_q)))
+	cache.logger.Info("cache hit", zap.String("key", KeyEncode(align_q, SimpleKeyPrefix)))
 	cache.cacheMetrics.CacheMetricsHit(*res[0])
 	return *res[0], nil
 }
+
+// Splits a result into buckets as defined by {buckets} and sets them in Redis
+// Buckets are assumed to be in order + non-overlapping
+func (cache *RedisCache) SetAsBuckets(result *storage.PromResult, buckets []*storage.FetchQuery) {
+	results := make([]*storage.PromResult, len(buckets))
+	for i := range buckets {
+		results[i] = createEmptyPromResult()
+	}
+	for _, ts := range result.PromResult.Timeseries {
+		idx := 0
+		for i, b := range buckets {
+			start := b.Start.Unix()
+			end := b.End.Unix()
+			prev := idx
+			// Convert millisecond timestamp to second timestamp
+			// Get all samples within the bucket
+			for idx < len(ts.Samples) && ts.Samples[idx].Timestamp/1000 >= start && ts.Samples[idx].Timestamp/1000 < end {
+				idx += 1
+			}
+			// Nothing to add
+			if prev == idx {
+				continue
+			}
+			results[i].PromResult.Timeseries = append(results[i].PromResult.Timeseries,
+				&prompb.TimeSeries{
+					Labels:  ts.Labels,
+					Samples: ts.Samples[prev:idx],
+				},
+			)
+			// No need to continue
+			if idx == len(ts.Samples) {
+				break
+			}
+		}
+	}
+	cache.Set(buckets, results, BucketKeyPrefix)
+}
+
+// Function that checks whether the query data is in Redis using a sliding window
+// If it is not, it gets the result from M3DB
+//
+// Currently, this is done by splitting data into buckets
+// and checking if we have enough buckets that we can retrieve from Redis to recreate the result
+func BucketWindowGetOrFetch(
```

Comment:
Can we put a latency metric on these Redis methods?

Reply:
added